File: packages/metadata_core/test/provenance_test.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
import 'package:metadata_core/metadata_core.dart';
import 'package:test/test.dart';

void main() {
  group('parseProvenance', () {
    test('maps the four known names', () {
      expect(parseProvenance('exif'), Provenance.exif);
      expect(parseProvenance('container'), Provenance.container);
      expect(parseProvenance('filesystem'), Provenance.filesystem);
      expect(parseProvenance('unknown'), Provenance.unknown);
    });

    test('falls back to unknown on anything else', () {
      expect(parseProvenance('EXIF'), Provenance.unknown);
      expect(parseProvenance(''), Provenance.unknown);
      expect(parseProvenance(null), Provenance.unknown);
      expect(parseProvenance(42), Provenance.unknown);
      expect(parseProvenance('made-up-source'), Provenance.unknown);
    });
  });

  group('MetaField', () {
    test('carries value with provenance', () {
      const MetaField<String> field =
          MetaField<String>('2026:09:14 11:05:23', Provenance.exif);
      expect(field.value, '2026:09:14 11:05:23');
      expect(field.provenance, Provenance.exif);
      expect(field.isKnown, isTrue);
    });

    test('unknown provenance is never known', () {
      const MetaField<String> field =
          MetaField<String>('x', Provenance.unknown);
      expect(field.isKnown, isFalse);
      const MetaField<String> empty =
          MetaField<String>(null, Provenance.unknown);
      expect(empty.isKnown, isFalse);
    });
  });
}

File: native/vision-core/src/test/kotlin/com/visionengine/core/ProvenanceTest.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================
package com.visionengine.core

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ProvenanceTest {
    @Test
    fun mapsKnownNames() {
        assertEquals(Provenance.exif, parseProvenance("exif"))
        assertEquals(Provenance.container, parseProvenance("container"))
        assertEquals(Provenance.filesystem, parseProvenance("filesystem"))
        assertEquals(Provenance.unknown, parseProvenance("unknown"))
    }

    @Test
    fun fallsBackToUnknown() {
        assertEquals(Provenance.unknown, parseProvenance("EXIF"))
        assertEquals(Provenance.unknown, parseProvenance(""))
        assertEquals(Provenance.unknown, parseProvenance(null))
        assertEquals(Provenance.unknown, parseProvenance(42))
        assertEquals(Provenance.unknown, parseProvenance("made-up-source"))
    }
}

class MetaFieldTest {
    @Test
    fun carriesValueWithProvenance() {
        val field = MetaField("2026:09:14 11:05:23", Provenance.exif)
        assertEquals("2026:09:14 11:05:23", field.value)
        assertEquals(Provenance.exif, field.provenance)
        assertTrue(field.isKnown)
    }

    @Test
    fun unknownProvenanceIsNeverKnown() {
        assertTrue(!MetaField("x", Provenance.unknown).isKnown)
        assertTrue(!MetaField<String>(null, Provenance.unknown).isKnown)
        assertTrue(!MetaField<String>(null, Provenance.exif).isKnown)
    }
}
